### PR TITLE
Refactoring of write methods, adding device selection

### PR DESCRIPTION
**Pull request by bensartori**
With this refactoring the handling of different write methods (up to now libusb and hidapi) has moved to an own class and its sub classes. This is a preparation to smoothly add bluetooth as a third write method in future and to be able to implement the following.

Additionally, there is the possibility list the connected devices and choose the one to be programmed. This way multiple devices may be connected to one computer and be programmed individually. With this change the endpoint number is not hardcoded anymore and the newer types of the name badge (with other USB chips) are found and programmed.

See command line options -M and -H / LedNameBadge.write(), LedNameBadge.get_available_methods(), LedNameBadge.get_available_device_ids() / class WriteMethod and its two sub classes.

All (ok, most) the decisions, validations and messages on choosing write mehtods and device ids are now concentrated in LedNameBadge._find_write_method().

It should still run on Windows (tested with pPython 3), MacOs (untested) and Python 2 (tested with Ubuntu linux)

Draft: some method documentation is still missing.

### CODE REVIEW

File: lednamebadge.py [hunk 1]
```diff
@@ -523,84 +802,267 @@ def header(lengths, speeds, modes, blinks, ants, brightness=100, date=datetime.n
...
-                    dev_info[0].manufacturer_string, dev_info[0].product_string, dev_info[0].interface_number, dev_info[0].usage_page))
-            else:
-                print("No led tag with vendorID 0x0416 and productID 0x5020 found.")
-                print("Connect the led tag and run this tool as root.")
+        if method == 'auto':
+            if sys.version_info[0] < 3:
+                method = libusb.get_name()
+                print("Preferring method %s over %s with Python 2.x" % (libusb.get_name(), hidapi.get_name()))
+                print("because of https://github.com/jnweiger/led-badge-ls32/issues/9")
+            elif sys.platform.startswith('darwin'):
```
Comment (bensartori) [inline lednamebadge.py:877]: Hello @s-light, Jürgen gave me the info, that you might be able to test this new version with MacOs? That would be great!

Comment (s-light) [reply lednamebadge.py:877]: oh sorry - not really..
i have no led-badge myself - and the macos devices are owned by my mother & brother - currently not living near by...

Comment (bensartori) [reply lednamebadge.py:877]: Ok, thanks and sorry for bothering...

Comment (s-light) [reply lednamebadge.py:877]: all fine :-)

File: README.md [hunk 2]
```diff
@@ -320,9 +427,10 @@ Just run `plantuml "*.puml"` from the `photos` directory to regenerate all diagr
...
 ## Related References (for USB-Serial devices)
- * https://github.com/Caerbannog/led-mini-board
- * http://zunkworks.com/projects/programmablelednamebadges/
- * https://github.com/DirkReiners/LEDBadgeProgrammer
- * https://bitbucket.org/bartj/led/src
- * http://www.daveakerman.com/?p=1440
- * https://github.com/stoggi/ledbadge
+
+* https://github.com/Caerbannog/led-mini-board
+* http://zunkworks.com/projects/programmablelednamebadges/
```
Comment (jnweiger) [inline README.md:432]:
```suggestion
```

That site no longer exists.

**Review by jnweiger — APPROVED** [review]
Thanks for the refactor, documentation, tests!

Two minor nitpicks:
* tests/abstract_witre_method_test.py is a typo -> _write_
* zunkworks.com seems gone.

**Comment by bensartori** [comment]
Mega! ;) Thanks.

**Pull request merged**